_posts/2025-09-19-minimizing-dependencies.md

```diff
@@ -0,0 +1,122 @@
+potential risks. Engineers have deliberately and carefully worked to reduce
+dependencies, replacing them with safer, purpose-built code or eliminating them
+entirely.
+
+## Risk of Dependencies
+
+While using pre-existing libraries can speed up development and productivity, in
+Bitcoin Core, securitiy is instead prioritized. Relying on dependencies can
+create security risks in several ways:
+
+- **Security vulnerabilities**: External libraries can contain bugs that expose
+  Bitcoin Core to crashes, remote code execution, or other exploits.
+- **Consensus fragility**: Subtle differences between library versions or
+  platforms can lead to consensus splits. Satoshi’s early use of OpenSSL
+  signature parsing nearly caused just such a problem.
+- **Supply-chain risk**: If an upstream project ships a malicious version,
+  whether by coercion, compromise, or mistake, Bitcoin Core could be exposed to
+  a backdoor.
+- **Build reproducibility**: Reproducible builds, critical for verifiability,
+  are harder to achieve when relying on third-party projects.
+- **Maintenance burden**: Developers must track upstream changes, apply security
+  patches, and ensure cross-platform compatibility.
+
+Every dependency adds to the attack surface. Reducing them increases resilience.
+
+## Removing Dependencies
+
+Over time Bitcoin Core has steadily removed unnecessary and risky dependencies,
+deleting them entirely, or replacing them with Bitcoin-specific, minimal
+implementations. Here are the most significant removals:
+
+- **OpenSSL (replaced with libsecp256k1)**
+  - OpenSSL was used for ECDSA, RNG, and X.509 parsing. Its complexity and
+    vulnerabilities (e.g. Heartbleed) were dangerous.
+  - Replaced by [libsecp256k1][], a purpose-built library, developed by Bitcoin
+    Core contributors.
+  - [Bitcoin Core #6954][bc 6954] (2016).
+
+- **Protobuf / BIP70 Payment Protocol**
+  - Bitcoin Core once supported BIP70, which required Protobuf and OpenSSL’s
+    X.509 certificate handling.
+  - This added attack surface and complexity for a feature rarely used.
+  - Fully removed in [Bitcoin Core #17165][bc 17165] (v0.20.0).
+
+- **Berkeley DB (BDB)**
+  - Originally the wallet backend.
+  - Known for brittle lock limits.
+  - Replaced by SQLite for descriptor wallets.
+
+- **miniupnpc and libnatpmp**
+  - Used for automatic NAT traversal and port forwarding.
+  - Both libraries had recurring security issues and were unreliable in
+    practice.
+  - miniUPnP removed in [Bitcoin Core #31130][bc 31130] (v29.0).
+  - libnatpmp removed in [Bitcoin Core #30043][bc 30043] (v29.0), replaced with
+    a minimal in-tree NAT implementation.
+
+- **Boost**
+  - Early versions of Core depended on large portions of Boost (filesystem,
+    signals, threads).
+  - Over the years, these were replaced with C++ standard library equivalents.
+  - Today, Boost is header-only and minimal, with plans to remove it entirely.
```

Comment:
> is header-only

Could link to https://github.com/bitcoin/bitcoin/pull/24301.

Reply:
Added